apps/aecore/lib/aecore/oracle/oracle_state_tree.ex

```diff
@@ -0,0 +1,185 @@
+  def insert_oracle(tree, oracle) do
+    add_oracle(tree, oracle, :insert)
+  end
+
+  @spec get_oracle(oracles_state(), binary()) :: map()
+  def get_oracle(tree, key) do
+    get(tree.oracle_tree, key)
+  end
+
+  @spec lookup_oracle?(oracles_state(), binary()) :: boolean()
+  def lookup_oracle?(tree, key) do
+    lookup?(tree, key, :oracle)
+  end
+
+  @spec enter_query(oracles_state(), map()) :: oracles_state()
+  def enter_query(tree, query) do
+    add_query(tree, query, :enter)
+  end
+
+  @spec insert_query(oracles_state(), map()) :: oracles_state()
+  def insert_query(tree, query) do
+    add_query(tree, query, :insert)
+  end
+
+  @spec get_query(oracles_state(), binary()) :: map()
+  def get_query(tree, key) do
+    get(tree.oracle_tree, key)
+  end
+
+  @spec lookup_query?(oracles_state(), binary()) :: boolean()
+  def lookup_query?(tree, key) do
+    lookup?(tree, key, :oracle_query)
+  end
+
+  defp initialize_deletion(tree, expired_oracles) do
+    expired_cache = get_expired_cache_ids(tree.oracle_cache_tree, expired_oracles)
+
+    new_map_tree =
+      Enum.reduce(expired_oracles, tree, fn {account_pubkey, _expires}, acc_tree ->
+        %{acc_tree | oracle_tree: delete(acc_tree.oracle_tree, account_pubkey)}
+      end)
+
+    Enum.reduce(expired_cache, new_map_tree, fn cache, acc_tree ->
+      %{acc_tree | oracle_cache_tree: delete(acc_tree.oracle_cache_tree, cache)}
+    end)
+  end
+
+  defp add_oracle(tree, oracle, how) do
+    id = oracle.owner
+    expires = oracle.expires
+    serialized = Serialization.rlp_encode(oracle, :oracle)
+
+    new_oracle_tree =
+      case how do
+        :insert ->
+          insert(tree.oracle_tree, id, serialized)
+
+        :enter ->
+          enter(tree.oracle_tree, id, serialized)
+      end
+
+    new_oracle_cache_tree = cache_push(tree.oracle_cache_tree, {:oracle, id}, expires)
+    %{oracle_tree: new_oracle_tree, oracle_cache_tree: new_oracle_cache_tree}
+  end
+
+  defp add_query(tree, query, how) do
+    oracle_id = query.oracle_address
+
+    id =
+      OracleQueryTx.id(
+        query.sender_address,
+        query.sender_nonce,
+        oracle_id
+      )
+
+    tree_id = oracle_id <> id
+    expires = query.sender_address
+    serialized = Serialization.rlp_encode(query, :oracle_query)
+
+    new_oracle_tree =
+      case how do
+        :insert ->
+          insert(tree.oracle_tree, tree_id, serialized)
+
+        :enter ->
+          enter(tree.oracle_tree, tree_id, serialized)
+      end
+
+    new_oracle_cache_tree = cache_push(tree.oracle_cache_tree, {:query, oracle_id, id}, expires)
+    %{oracle_tree: new_oracle_tree, oracle_cache_tree: new_oracle_cache_tree}
+  end
+
+  defp insert(tree, key, value) do
+    PatriciaMerkleTree.enter(tree, key, value)
+  end
+
+  defp enter(tree, key, value) do
+    PatriciaMerkleTree.enter(tree, key, value)
+  end
+
+  defp delete(tree, key) do
+    PatriciaMerkleTree.delete(tree, key)
+  end
+
+  defp lookup?(tree, key, where) do
```

Comment:
lookup. where and which_tree sound weird as names

Reply:
@d-velev 
Yes, I'm agree about the incorrectness of the function name ```lookup```, it should be ```exists```. After all it checks whether the obj under the key exists. We should correct this.

Since the existence of more than one Trie, the function is intended to work more generic way. To give the ability to ease the distinction from ```WHERE``` you want to make the selection and the function's name ```which_tree``` speaks for itself.